streamingccl: support ingestion job to pause on error

gh-casper opened the pull request:
Previously, an ingestion job fails on error and the tenant state is
kept after failure. The user is supposed to execute another
RESTORE FROM REPLICATION STREAM statement to resume ingestion from
a given checkpoint.

Now we introduce the new semanics: the job will be paused on error,
the user can RESUME JOB to start from the previous checkpoint or
cancel the job to roll back (in another PR).

We also drop AS OF syntax as it conflicts with the purpose of pause
on error. AS OF assumes users resume ingesion for an existent tenant
from a given checkpoint and creates a new ingestion job while the
previous ingestion job is paused on error, making a bad user
experience.

Closes: https://github.com/cockroachdb/cockroach/issues/83415

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/83112)
<!-- Reviewable:end -->

gh-casper reviewed (commented):
Quickly flushing out my comment on removing AS OF. Other revision and comments come later.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83112#-:-N57lUYq9VYaeBiaspG4:b-vy92l9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gh-casper, @miretskiy, and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83112#-N52ManzS2iFc0EoaRh3-r2-161:-N57iTVNBo-SyOo3ajcG:bz05aeb) ([raw file](https://github.com/cockroachdb/cockroach/blob/975e8b8f2ac27e60dfc25aa8d04cc4074e34b8f3/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L161)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

It seems reasonable to save the start time of the most recent resumption like you've done here, but I'm not seeing why that requires giving up still having the overall StartTime of the job.

We could leave AS OF in place and keep the start time as part of the initial state.  On resume, if we have a non-empty high watermark, we use that as our startTime. Otherwise, we use the start time from the initial state.
</blockquote></details>

I think in the UX meeting, even though syntax is not determined yet, we have agreed on dropping `AS OF`. This is very related to the desired customer journey. 

Users should always start stream ingestion with initial scan. Because, in the end, this statement will check if tenant exists or not, create the tenant in offline mode, start ingestion, pause on error, resume, cutover, make the tenant online. We can't ingest into an existent tenant with half state, which is all this `AS OF` about. An offline existent tenant with half state means an ingestion job get paused and not rolled back yet, we don't want to create ingestion job to touch this tenant again.

Regarding why we want to change it in this PR, well, it's relevant.  AS OF kinda conflicts the purpose of pausing on error for the reason above.
___
*[`pkg/sql/parser/sql.y` line 3182 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83112#-N52Vu7Ihe1FBtZ-PUpD-r2-3182:-N57lNvAAP5KMhuBr1xF:b55d52s) ([raw file](https://github.com/cockroachdb/cockroach/blob/975e8b8f2ac27e60dfc25aa8d04cc4074e34b8f3/pkg/sql/parser/sql.y#L3182)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I will second @stevendanna question: why do we need to change this?
Why change this as part of this PR?
</blockquote></details>

See my other comment with Steven. 


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83112)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gh-casper, @miretskiy, and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83112#-N52ManzS2iFc0EoaRh3-r2-161:-N57u91JBfhcak5ydoMt:b-v1ykpo) ([raw file](https://github.com/cockroachdb/cockroach/blob/975e8b8f2ac27e60dfc25aa8d04cc4074e34b8f3/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L161)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

I think in the UX meeting, even though syntax is not determined yet, we have agreed on dropping `AS OF`. This is very related to the desired customer journey. 

Users should always start stream ingestion with initial scan. Because, in the end, this statement will check if tenant exists or not, create the tenant in offline mode, start ingestion, pause on error, resume, cutover, make the tenant online. We can't ingest into an existent tenant with half state, which is all this `AS OF` about. An offline existent tenant with half state means an ingestion job get paused and not rolled back yet, we don't want to create ingestion job to touch this tenant again.

Regarding why we want to change it in this PR, well, it's relevant.  AS OF kinda conflicts the purpose of pausing on error for the reason above.

</blockquote></details>

Thanks for the explanation.  Apologies I've not engaged with that UX document as much as I should have yet.

In the behaviour we are driving towards here, all errors would result in a PAUSED rather than a running job. And thus users respond to that failure via RESUME rather than by creating a new job. And, all changes to a job would happen via some combination of PAUSE and ALTER.  I agree that in that world, there are very few cases where we want to _start_ a job with an explicit StartTime.

That said, it makes me a little nervous not having it around as a tool in case the job has failed in a way that we don't expect (or has a problem we can't fix with ALTER).

I suppose it is easy enough to add back if we decide we do need it down the road.

I'll leave this one to your discretion.


<!-- Sent from Reviewable.io -->

gh-casper reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83112)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 144 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83112#-N52ManzS2iFc0EoaRh3-r2-144:-N5bhADV4zkcS3HYPBDm:b-tqkaml) ([raw file](https://github.com/cockroachdb/cockroach/blob/975e8b8f2ac27e60dfc25aa8d04cc4074e34b8f3/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L144)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Should we log here?
</blockquote></details>

Added a few logs.



_Code quote:_
```Go
status.StreamStatus != streampb.StreamReplicationStatus_UNKNOWN_STREAM_STATUS_R
```
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83112#-N52ManzS2iFc0EoaRh3-r2-161:-N5beuuzDixS7g9XaIKD:bs8vo43) ([raw file](https://github.com/cockroachdb/cockroach/blob/975e8b8f2ac27e60dfc25aa8d04cc4074e34b8f3/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L161)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Thanks for the explanation.  Apologies I've not engaged with that UX document as much as I should have yet.

In the behaviour we are driving towards here, all errors would result in a PAUSED rather than a running job. And thus users respond to that failure via RESUME rather than by creating a new job. And, all changes to a job would happen via some combination of PAUSE and ALTER.  I agree that in that world, there are very few cases where we want to _start_ a job with an explicit StartTime.

That said, it makes me a little nervous not having it around as a tool in case the job has failed in a way that we don't expect (or has a problem we can't fix with ALTER).

I suppose it is easy enough to add back if we decide we do need it down the road.

I'll leave this one to your discretion.


</blockquote></details>

As discussed offline, we will drop AS OF SYSTEM TIME.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 211 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83112#-N52ManzS2iFc0EoaRh3-r2-211:-N5bfjde2lFJeUJHzwr6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/975e8b8f2ac27e60dfc25aa8d04cc4074e34b8f3/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L211)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] Can we avoid having two slightly different messages here:

```suggestion
    errorMessage := fmt.Sprintf("ingestion job failed (%v), requesting pause", err)
		log.Warningf(resumeCtx, errMessage)	
```
</blockquote></details>

Done.



_Code quote:_
```Go
ingest(resumeCtx, p, s.job)
```
___
*[`pkg/ccl/streamingccl/streampb/stream.proto` line 127 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83112#-N52ManzS2iFc0EoaRh7-r2-127:-N5bfuko3SVqsHsKbIS3:b-ta1x61) ([raw file](https://github.com/cockroachdb/cockroach/blob/975e8b8f2ac27e60dfc25aa8d04cc4074e34b8f3/pkg/ccl/streamingccl/streampb/stream.proto#L127)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This reads a bit strange...
Also.. this sounds like a low_watermark and not a high watermark?
Perhaps avoid confusion and rename the field? Min timestamp or some such?
</blockquote></details>

Will fix that in the other PR.


<!-- Sent from Reviewable.io -->

gh-casper committed:
streamingccl: support ingestion job to pause on error

Previously, an ingestion job fails on error and the tenant state is
kept after failure. The user is supposed to execute another
RESTORE FROM REPLICATION STREAM statement to resume ingestion from
a given checkpoint.

Now we introduce the new semanics: the job will be paused on error,
the user can RESUME JOB to start from the previous checkpoint or
cancel the job to roll back (in another PR).

We also drop AS OF syntax as it conflicts with the purpose of pause
on error. AS OF assumes users resume ingesion for an existent tenant
from a given checkpoint and creates a new ingestion job while the
previous ingestion job is paused on error, making a bad user
experience.

Release note: None

stevendanna commented on pkg/ccl/streamingccl/streamproducer/stream_lifetime.go:
```diff
@@ -123,11 +123,14 @@ func updateReplicationStreamProgress(
 
-			if shouldUpdatePTS := ptsRecord.Timestamp.Less(ts); shouldUpdatePTS {
-				if err = ptsProvider.UpdateTimestamp(ctx, txn, ptsID, ts); err != nil {
+			// TODO(casper): Error out when the protected timestamp moves backward as the ingestion
+			// processors may consume kv changes that are not protected. We are fine for now
+			// for the sake of long GC window.
```
Is this something we've seen.  Can we describe how it happens?

stevendanna commented on pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go:
```diff
@@ -350,4 +314,104 @@ func TestTenantStreamingProducerJobTimedOut(t *testing.T) {
+	pausedCheckpoint := streamIngestionStats(t, c.destSysSQL, ingestionJobID).
+		ReplicationLagInfo.MinIngestedTimestamp
+	// Check we paused at a timestamp greater than the previously reached high watermark
+	require.True(t, srcTime.Before(pausedCheckpoint.GoTime()))
```
I wonder if this will be flaky.  Is there anything that guarantees that we actually advance the checkpoint?  Perhaps should should be a Before or Equal check.

stevendanna commented on pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go:
```diff
@@ -350,4 +314,104 @@ func TestTenantStreamingProducerJobTimedOut(t *testing.T) {
+
+	var srcTime time.Time
+	c.srcSysSQL.QueryRow(t, "SELECT clock_timestamp()").Scan(&srcTime)
+	c.waitUntilHighWatermark(srcTime, jobspb.JobID(ingestionJobID))
```
I believe the TestServerInerface has a Clock() method that you could use to get the srcTime directly.

stevendanna reviewed (commented):
Let's get this merged. I've left some nitpic on the log messages that you can take or leave at your discretion.

Can you add the reasoning for the removal of AS OF SYSTEM TIME to the commit message?

gh-casper reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83112)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 144 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83112#-N52ManzS2iFc0EoaRh3-r3-144:-N6K9keDDyvcjQbQrkSk:b-x76yjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7d15e5f223be8342ba31d0cfabab932fb32926/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L144)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Potentialy mentioned in the other PR: Do we definitely want to return here, as opposed to continuing to retry on error?
</blockquote></details>

I'll have another PR to implement retry around `ingest()`:
`for r := retry.Start(ro); r.Next(); {
     err := ingest()
     if errors.isPermernantJobError(err) {
         break
     }
}`

This issue is tracked here: https://github.com/cockroachdb/cockroach/issues/83450.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 151 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83112#-N52ManzS2iFc0EoaRh3-r3-151:-N6KBUcTA9T9lbQXVt39:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7d15e5f223be8342ba31d0cfabab932fb32926/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L151)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Is the status actually unknown? How about something like

```suggestion
				"producer job %d has status %v, retrying", status.StreamStatus, streamID)
```
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 159 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83112#-N52ManzS2iFc0EoaRh3-r3-159:-N6KBcrXCtenbAqV_L0j:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7d15e5f223be8342ba31d0cfabab932fb32926/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L159)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

```suggestion
		log.Infof(ctx,
			"producer job %d is active, creating a stream replication plan", streamID)
```
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamproducer/stream_lifetime.go` line 128 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83112#-N52ManzS2iFc0EoaRhG-r3-128:-N6KDgBA4ba3Fhy5cRD8:b-q5ahw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7d15e5f223be8342ba31d0cfabab932fb32926/pkg/ccl/streamingccl/streamproducer/stream_lifetime.go#L128)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Is this something we've seen.  Can we describe how it happens?
</blockquote></details>

Added a detailed description for why this happens here.



_Code quote:_
```Go
TODO(casper): Error out when th
```
___
*[`pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go` line 339 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83112#-N52ManzS2iFc0EoaRh6-r3-339:-N6MG7nS-76rKvbGBRS7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7d15e5f223be8342ba31d0cfabab932fb32926/pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go#L339)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I believe the TestServerInerface has a Clock() method that you could use to get the srcTime directly.
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go` line 350 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83112#-N52ManzS2iFc0EoaRh6-r3-350:-N6MG6k_7a1ZQ2SBQ6Sg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7d15e5f223be8342ba31d0cfabab932fb32926/pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go#L350)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I wonder if this will be flaky.  Is there anything that guarantees that we actually advance the checkpoint?  Perhaps should should be a Before or Equal check.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (approved):
Thanks for addressing some of the feedback and fixing up the commit message.

Removing the AS OF SYSTEM TIME without a good replacement makes me a little nervous. But we'll hopefully be addressing that in short order.

We should file an issue for the protected timestamp issue if we don't already have one.

Not sure if @miretskiy wants to have another look.

miretskiy commented:
Let's get this in.

gh-casper commented:
bors r+

gh-casper commented:
Filed an issue for the protected timestamp issue: https://github.com/cockroachdb/cockroach/issues/84086

gh-casper commented:
bors r+

gh-casper commented:
bors r+

gh-casper commented:
bors r+
